pkg/combustion/kubernetes.go

```diff
@@ -122,28 +138,116 @@ func configureRKE2(ctx *image.Context) (string, error) {
+	auditDefaultCNI := func() {
+		auditMessage := fmt.Sprintf("Kubernetes CNI not explicitly set, defaulting to: %s", cniDefaultValue)
+		log.Audit(auditMessage)
+	}
+
+	config := map[string]any{}
+
 	if !isComponentConfigured(ctx, k8sConfigDir) {
-		zap.S().Info("Kubernetes config file not provided")
-		return "", nil
+		auditDefaultCNI()
+		zap.S().Infof("Kubernetes config file not provided, proceeding with CNI: %s", cniDefaultValue)
+
+		config[cniKey] = cniDefaultValue
+		return config, nil
 	}
 
 	configDir := generateComponentPath(ctx, k8sConfigDir)
 	configFile := filepath.Join(configDir, k8sConfigFile)
 
-	_, err := os.Stat(configFile)
+	b, err := os.ReadFile(configFile)
 	if err != nil {
 		if errors.Is(err, fs.ErrNotExist) {
-			return "", fmt.Errorf("kubernetes component directory exists but does not contain config.yaml")
+			return nil, fmt.Errorf("kubernetes component directory exists but does not contain config.yaml")
+		}
+		return nil, fmt.Errorf("reading kubernetes config file: %w", err)
+	}
+
+	if err = yaml.Unmarshal(b, &config); err != nil {
+		return nil, fmt.Errorf("parsing kubernetes config file: %w", err)
+	}
+
+	if _, ok := config[cniKey]; !ok {
+		auditDefaultCNI()
+		zap.S().Infof("CNI not set in config file, proceeding with CNI: %s", cniDefaultValue)
+
+		config[cniKey] = cniDefaultValue
+	}
+
+	return config, nil
+}
+
+func extractCNI(config map[string]any) (cni string, multusEnabled bool, err error) {
+	switch configuredCNI := config[cniKey].(type) {
+	case string:
+		if configuredCNI == "" {
+			return "", false, fmt.Errorf("cni not configured")
 		}
-		return "", fmt.Errorf("error checking kubernetes config file: %w", err)
+
+		var cnis []string
+		for _, cni = range strings.Split(configuredCNI, ",") {
+			cnis = append(cnis, strings.TrimSpace(cni))
+		}
+
+		return parseCNIs(cnis)
+
+	case []string:
+		return parseCNIs(configuredCNI)
+
+	case []any:
+		var cnis []string
+		for _, cni := range configuredCNI {
+			c, ok := cni.(string)
+			if !ok {
+				return "", false, fmt.Errorf("invalid cni value: %v", cni)
+			}
+			cnis = append(cnis, c)
+		}
+
+		return parseCNIs(cnis)
+
+	default:
+		return "", false, fmt.Errorf("invalid cni: %v", configuredCNI)
+	}
+}
+
+func parseCNIs(cnis []string) (cni string, multusEnabled bool, err error) {
+	const multusPlugin = "multus"
+
+	switch len(cnis) {
+	case 1:
+		cni = cnis[0]
+		if cni == multusPlugin {
+			return "", false, fmt.Errorf("multus must be used alongside another primary cni selection")
+		}
+	case 2:
+		if cnis[0] == multusPlugin {
+			cni = cnis[1]
+			multusEnabled = true
+		} else {
+			return "", false, fmt.Errorf("multiple cni values are only allowed if multus is the first one")
+		}
+	default:
+		return "", false, fmt.Errorf("invalid cni value: %v", cnis)
+	}
+
+	return cni, multusEnabled, nil
+}
+
+func storeKubernetesConfig(ctx *image.Context, config map[string]any, distribution string) (string, error) {
+	data, err := yaml.Marshal(config)
```

Comment:
What made you change the user-specified config by explicitly entering the default? I could see an argument for doing this, just curious what the rationale was. Feels a bit wrong to modify their files, so I at least want to be able to say we've had the discussion on why it's needed.

Reply:
If a user does not provide a configuration file or provides one without a `cni` configuration,  we have to modify the configuration before saving it. The default one in those cases would be Canal, however, @rdoxenham suggested that for Edge deployments we should always go for Cilium if something else is not explicitly set.

Reply:
Whilst we're trying to be flexible, we should also consider being quite prescriptive with our desired configuration that customers are using. Cilium is a strong choice here and if we can standardise on this as the default, I think that makes sense.

Reply:
Works for me 👍🏼